### PR TITLE
Python script to generate lists of code scanning queries in CSV format

### PR DESCRIPTION
**Pull request by sj**
tl;dr @jhutchings1 started working on a PowerShell script that would generate a CSV file containing metadata for all queries in code scanning query packs. This is a Python translation of that script. See https://github.com/github/codeql/pull/3664 for the PowerShell script.

Back in https://github.com/github/codeql/pull/3664, I [pointed out some issues](https://github.com/github/codeql/pull/3664#issuecomment-660112909) with that particular approach using PowerShell:
 - The CSV file is manually generated and is therefore malformed due to not properly escaping/removing newlines and commas from the query metadata
 - Very few people in the CodeQL engineering team run Windows, and [running PowerShell on Ubuntu is challenging](https://github.com/PowerShell/PowerShell/issues/12626). Having this functionality in PowerShell means it likely won't be maintained and would require rewriting soon.

I therefore embarked on [translating the script to Python back in mid-July](https://github.com/github/codeql/pull/3664#issuecomment-660124671). That all worked, but we never actually got to merge it. There were some issues with getting it to run on Actions, but that's all resolved now.

On the long term, we might want to move functionality like this into the CodeQL CLI (for one thing, the current approach is *very* slow), but the code scanning team are under some pressure to make this data available now (not in the least so it can be used for documentation purposes). I therefore suggest that we consider this script (and Actions workflow) an MVP for us to learn from: if this functionality turns out to be useful enough to have it in the CodeQL CLI, then we can do that later.

For ease of review: [this is the (zipped) CSV data](https://github.com/github/codeql/files/5155969/code-scanning-query-lists.zip) generated by the script, and [here's the same data imported into Google Sheets](https://docs.google.com/spreadsheets/d/11ZPQwvI-pH5Ggyxek5l93w1gLaWCX3QRMgKGQGbPvPw).

### CODE REVIEW

**Comment by sj** [comment]
@alexet or @hmakholm: could either of you take a quick look please?

cc @adityasharad for info

**Comment by sj** [comment]
Question for @jhutchings1: how often do you need an up-to-date version of this list? As it stands, this job runs on every push to every branch, but I think it'd be fine to change that to only run on pushes to `main`. Let me know what you think!

File: .github/workflows/query-list.yml [hunk 1]
```diff
@@ -0,0 +1,39 @@
+name: Build code scanning query list
+
+on:
+  push:
```
Comment (adityasharad) [inline .github/workflows/query-list.yml:4]: I agree with your thinking that this should only be pushes to `main`, and perhaps to the RC branches.

File: .github/workflows/query-list.yml [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+    runs-on: ubuntu-latest
+
+    steps:
+    - name: Clone self (github/codeql)
+      uses: actions/checkout@v2
+      with:
+        path: codeql 
+    - name: Clone github/codeql-go
+      run: git clone https://github.com/github/codeql-go
```
Comment (adityasharad) [inline .github/workflows/query-list.yml:17]: You can use `actions/checkout` again here, just specify `path: codeql-go`. The advantage is that it reduces the fetch depth.

File: .github/workflows/query-list.yml [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+    - name: Set up Python 3.8
+      uses: actions/setup-python@v2
+      with:
+        python-version: 3.8
+    - name: Download CodeQL CLI
+      uses: dsaltares/fetch-gh-release-asset@master
```
Comment (adityasharad) [inline .github/workflows/query-list.yml:23]: My preference would be to avoid a dependency on an external Action, and do this by directly calling https://developer.github.com/v3/repos/releases/#get-a-release-asset.

Bash step equivalent:
```
env:
  GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
run: |
  DOWNLOAD_URL=$(gh api -X GET repos/github/codeql-cli-binaries/releases/latest | jq -r '.assets | map(select(.name == "codeql-linux64.zip"))[0].url')
  gh api -X GET -H accept:application/octet-stream "$DOWNLOAD_URL" > codeql-linux64.zip
```

If you wish to stick with the external Action, please pin the Action version to a specific full-length SHA, and do not use a branch reference like `master`.

File: misc/scripts/generate-code-scanning-query-list.py [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+# Define which languages and query packs to consider
+languages = [ "cpp", "csharp", "go", "java", "javascript", "python"]
+packs = [ "code-scanning", "security-and-quality", "security-extended" ]
+
+# Define CodeQL search path so it'll find the CodeQL repositories:
+#  - anywhere in the current Git clone (including current working directory)
+#  - the 'codeql' subdirectory of the cwd
+#
+# (and assumes the codeql-go repo is in a similar location)
+codeql_search_path = "./codeql:./codeql-go:"   # will be extended further down
```
Comment (adityasharad) [inline misc/scripts/generate-code-scanning-query-list.py:28]: @hmakholm may be able to advise whether you actually need to set the search path, if the CLI and the query repos are in the right sibling directories.

File: misc/scripts/generate-code-scanning-query-list.py [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+        ql/java/ql/src/MyQuery.ql
+    
+    If the filename is not part of a Git repo, the return value is the
+    same as the input value: the whole path.
+    """
+    dirname = os.path.dirname(filename)
+
+    git_toplevel_dir_subp = subprocess.run(
+        "git -C '%s' rev-parse --show-toplevel"  % dirname,
+        shell=True, capture_output=True, text=True
```
Comment (adityasharad) [inline misc/scripts/generate-code-scanning-query-list.py:52]: Do you need `shell=True`? That's a risk with an externally-controlled directory name.

File: misc/scripts/generate-code-scanning-query-list.py [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+    first_sha = subprocess.run(
+        "git -C '%s' rev-list --max-parents=0 HEAD" % dirname,
+        shell=True, capture_output=True, text=True
+    ).stdout.strip()
+
+    if first_sha == "b55526aa58a5b500fe5d1be2c7edd09075711d09": prefix = "github/codeql"
+    elif first_sha == "d14eb855fc88e086b587f6c9695b59eb230c79e7": prefix = "github/codeql-go"
+    else: prefix = os.path.basename(git_toplevel_dir)
+
+    return os.path.join(prefix, filename[len(git_toplevel_dir)+1:])
```
Comment (adityasharad) [inline misc/scripts/generate-code-scanning-query-list.py:71]: Looking at the output from the most recent Actions run, this works fine for `github/codeql-go`, but for `github/codeql` the prefix is only `codeql` (ie the directory name and not the repo name), so you seem to be incorrectly hitting the `else` case.

I suspect the reason is that `actions/checkout` by default only fetches up to depth 1, for speed. Hence looking up the first commit is not going to work. It works here for `codeql-go` because you did a manual clone instead of using `actions/checkout`.

Do you have to match against the first commit this way? I think there are simpler alternatives, e.g. looking at the output of `git remote --get-url origin` or `gh repo view | head -n 1`.

File: misc/scripts/generate-code-scanning-query-list.py [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+
+def get_query_metadata(key, metadata, queryfile):
+    """Returns query metadata or prints a warning to stderr if a particular piece of metadata is not available."""
+    if key in metadata: return single_spaces(metadata[key])
+    query_id = metadata['id'] if 'id' in metadata else 'unknown'
+    print("Warning: no '%s' metadata for query with ID '%s' (%s)" % (key, query_id, queryfile), file=sys.stderr)
+    return ""
+
+def subprocess_run(cmd):
+    return subprocess.run(cmd, shell=True, capture_output=True, text=True, env=os.environ.copy())
```
Comment (adityasharad) [inline misc/scripts/generate-code-scanning-query-list.py:89]: May want to use `check_run`, and drop `shell=True`.

File: misc/scripts/generate-code-scanning-query-list.py (hunk 6 above)
Comment (sj) [reply misc/scripts/generate-code-scanning-query-list.py:67]: I've indeed now switched to using `git remote -v` as a heuristic for finding the remote!

File: misc/scripts/generate-code-scanning-query-list.py (hunk 5 above)
Comment (sj) [reply misc/scripts/generate-code-scanning-query-list.py]: Excellent point. I've removed that.

File: .github/workflows/query-list.yml (hunk 3 above)
Comment (sj) [reply .github/workflows/query-list.yml]: I've now specified the full SHA. I think using the external Action is fine for this 'MVP'; if we decide to offer this functionality in a different way (e.g. integrated into CodeQL CLI) in the future, then we can reconsider.

File: misc/scripts/generate-code-scanning-query-list.py (hunk 7 above)
Comment (sj) [reply misc/scripts/generate-code-scanning-query-list.py]: I think you meant `subprocess.check_call` (rather than `check_run`), but that has a downside: you can't inspect `stdout` afterwards. So I opted for the equivalent `check=True`.

File: misc/scripts/generate-code-scanning-query-list.py (hunk 4 above)
Comment (sj) [reply misc/scripts/generate-code-scanning-query-list.py:28]: I tried that, but that didn't seem to work

**Comment by sj** [comment]
Thanks for the thorough review, @adityasharad! I've addressed your comments and I think this is now ready for another look. Note that I've also set up the Action to no longer run on all pushes, so any changes to this branch won't immediately trigger a run of the script!

File: misc/scripts/generate-code-scanning-query-list.py [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+except Exception as e:
+    print("Error: couldn't invoke CodeQL CLI 'codeql'. Is it on the path? Aborting.", file=sys.stderr)
+    raise e
+
+# Define CodeQL search path so it'll find the CodeQL repositories:
+#  - anywhere in the current Git clone (including current working directory)
+#  - the 'codeql' subdirectory of the cwd
+#
+# (and assumes the codeql-go repo is in a similar location)
+codeql_search_path = "./codeql:./codeql-go:"   # will be extended further down
```
Comment (hmakholm) [inline misc/scripts/generate-code-scanning-query-list.py:102]:
```suggestion
codeql_search_path = "./codeql:./codeql-go:."   # will be extended further down
```
Not sure how well an _empty_ string for a directory will work; better make it `.`.

File: misc/scripts/generate-code-scanning-query-list.py (hunk 4 above)
Comment (hmakholm) [reply misc/scripts/generate-code-scanning-query-list.py]: The unzipping instruction puts the `codeql` binary as `codeql-cli/codeql/codeql` due to the clash between the `codeql` directory prefix inside `codeql-linux64.zip` and the name of the `github/codeql` repository. So the usual sibling heuristic will indeed not work -- the relevant directories are uncles not siblings.

For use in automation I think it does feel safer to build a search path explicitly than to rely on default "action at a distance" heuristics, so :+1: from me.

**Comment by jf205** [comment]
👋 just driving by to say that we might to publish these query lists the along side the CodeQL query help eventually--we're currently looking at creating a new process for publishing the query help so that we can plan to move off Confluence sooner rather than later (see https://github.com/github/semmle-docs/issues/93#issuecomment-687797541).

File: misc/scripts/generate-code-scanning-query-list.py [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+
+
+try: # Check for `git` on path
+    git_version = subprocess_run(["git","--version"])
+except Exception as e:
+    print("Error: couldn't invoke 'git'. Is it on the path? Aborting.", file=sys.stderr)
+    raise e
+
+try: # Check for `codeql` on path
+    codeql_version = subprocess_run(["codeql","--version"])
```
[the remark this answers is not in the document]
Comment (sj) [reply misc/scripts/generate-code-scanning-query-list.py:92]: We don't actually need the version; it's just checking whether CodeQL is available. I've removed the unused variable now.

File: misc/scripts/generate-code-scanning-query-list.py (hunk 8 above)
Comment (sj) [reply misc/scripts/generate-code-scanning-query-list.py]: Done! I've also re-enabled the workflow on this branch to double-check it works.

**Comment by sj** [comment]
@hmakholm: I've made some changes following your suggestions. Could you and/or @adityasharad have a final look please? Thasks!

**Review by hmakholm — APPROVED** [review]
Looks fair to me.

**Comment by sj** [comment]
Hold on, the Action is reporting:
```
Warning: couldn't find query pack 'code-scanning' for language 'go'. Do you have the right repositories in the right places (search path: './codeql:./codeql-go:.')?
Warning: couldn't find query pack 'security-and-quality' for language 'go'. Do you have the right repositories in the right places (search path: './codeql:./codeql-go:.')?
Warning: couldn't find query pack 'security-extended' for language 'go'. Do you have the right repositories in the right places (search path: './codeql:./codeql-go:.')?
```

Just debugging

File: .github/workflows/query-list.yml [hunk 10]
```diff
@@ -0,0 +1,46 @@
+name: Build code scanning query list
+
+on:
+  push:
+    branches:
+     - master
```
Comment (adityasharad) [inline .github/workflows/query-list.yml:6]:
```suggestion
     - main
```

File: .github/workflows/query-list.yml [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+     - 'rc/**'
+     - querylist-python
```
Comment (adityasharad) [inline .github/workflows/query-list.yml:8]: For future testing purposes, I suggest you add a
```
pull_request:
  paths:
    - '.github/workflows/query-list.yml'
```

**Review by adityasharad — COMMENTED** [review]
Looks good! Couple of minor suggestions on the Actions trigger.

File: .github/workflows/query-list.yml (hunk 11 above)
Comment (sj) [reply .github/workflows/query-list.yml:8]: Done! I've also included the path to the script itself.

**Comment by sj** [comment]
> Hold on, the Action is reporting:
> `...`
> Just debugging

To confirm: this problem was fixed by fc8f01e. I've attached the [most recently generated](https://github.com/github/codeql/runs/1092222429) query list: [code-scanning-query-list.zip](https://github.com/github/codeql/files/5200825/code-scanning-query-list.zip)

**Comment by sj** [comment]
Thanks @adityasharad, great suggestions. All implemented. Could you take another look please?
